### PR TITLE
Added snort3 & libdaq recipes

### PR DESCRIPTION
**Pull request by RNGesus-exe**

### CODE REVIEW

File: meta-networking/recipes-connectivity/daq/libdaq_git.bb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+DEPENDS = "libpcap libpcre libdnet bison-native"
+
+SRC_URI = "git://github.com/snort3/libdaq.git;protocol=https;branch=master"
+SRCREV = "c757aaae5a70d5bbb3239c5a16bca217a5aca651" 
+
+S = "${WORKDIR}/git"
+
+inherit autotools pkgconfig
+
+DISABLE_STATIC = ""
```
Comment (kraj) [inline meta-networking/recipes-connectivity/daq/libdaq_git.bb:18]: why is this needed ?

File: meta-networking/recipes-connectivity/daq/libdaq_git.bb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+SUMMARY = "The dump DAQ test the various inline mode features "
+HOMEPAGE = "http://www.snort.org"
+SECTION = "libs"
+LICENSE = "GPL-2.0-only"
+LIC_FILES_CHKSUM = "file://COPYING;md5=79258250506422d064560a7b95b2d53e"
+
+PARALLEL_MAKE = ""
```
Comment (kraj) [inline meta-networking/recipes-connectivity/daq/libdaq_git.bb:7]: why is this needed ?

File: meta-networking/recipes-connectivity/snort/snort3_git.bb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+DESCRIPTION = "snort3 - a free lightweight network intrusion detection system for UNIX and Windows."
+HOMEPAGE = "http://www.snort.org/"
+SECTION = "net"
+LICENSE = "GPL-2.0-only"
+LIC_FILES_CHKSUM = "file://COPYING;md5=78fa8ef966b48fbf9095e13cc92377c5"
+
+DEPENDS = "luajit libdaq libdnet flex hwloc libpcap libpcre zlib libtirpc"
+RDEPENDS:${PN} += "bash"
```
Comment (kraj) [inline meta-networking/recipes-connectivity/snort/snort3_git.bb:8]: Can we remove this direct dependency ? perhaps convert the shell to not need bash extentions or move the concerning scripts to package of its own.

File: meta-networking/recipes-connectivity/snort/snort3_git.bb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+SRC_URI = "git://github.com/snort3/snort3.git;protocol=https;branch=master"
+SRCREV = "e1760a8dbb829bb3fcf1a340ab6cc4bb80a47ecd" 
+
+inherit cmake pkgconfig
+
+S = "${WORKDIR}/git"
+
+EXTRA_OECONF = "\
+    --enable-static-daq \
```
Comment (kraj) [inline meta-networking/recipes-connectivity/snort/snort3_git.bb:18]: does this mean we get both shared and static libs? is the library under GPL-2.0 as well? if so then we should try to make static library generation as non-default.

File: meta-networking/recipes-connectivity/daq/libdaq_git.bb (hunk 1 above)
Comment (RNGesus-exe) [reply meta-networking/recipes-connectivity/daq/libdaq_git.bb:18]: Removed it

File: meta-networking/recipes-connectivity/daq/libdaq_git.bb (hunk 2 above)
Comment (RNGesus-exe) [reply meta-networking/recipes-connectivity/daq/libdaq_git.bb:7]: Removed this as well

File: meta-networking/recipes-connectivity/snort/snort3_git.bb (hunk 3 above)
Comment (RNGesus-exe) [reply meta-networking/recipes-connectivity/snort/snort3_git.bb:8]: I tried to remove this dependency by creating a patch for the bash script but there are a lot of changes to be made in this script and I don't need it for my use case.

File: meta-networking/recipes-connectivity/snort/snort3_git.bb (hunk 4 above)
Comment (RNGesus-exe) [reply meta-networking/recipes-connectivity/snort/snort3_git.bb:18]: Yes It is under GPL-2.0 [snort3-license](https://github.com/snort3/snort3/blob/master/LICENSE) similar to libdaq. I removed these configurations as it is not needed for my use case.

File: meta-networking/recipes-connectivity/snort/snort3_git.bb (hunk 3 above)
Comment (kraj) [reply meta-networking/recipes-connectivity/snort/snort3_git.bb:8]: OK, can be move the script to a package of its own. ? something like below

PACKAGES += "${PN}-scripts"

FILES:${PN}-scripts = "/path/to/script/name"

RDEPENDS:${PN}-scripts = "bash"

**Comment by kraj** [comment]
also make sure its only enabled for architectures supporting luajit otherwise we see errors like this 

```
12:09:02 luajit was skipped: incompatible with host riscv64-yoe-linux (not in COMPATIBLE_HOST)
12:09:10 ERROR: Nothing RPROVIDES 'snort3-dev' (but /mnt/jenkins/workspace/yocto-world-glibc/sources/meta-openembedded/meta-networking/recipes-connectivity/snort/snort3_git.bb RDEPENDS on or otherwise requires it)
12:09:10 No eligible RPROVIDERs exist for 'snort3-dev'
12:09:10 NOTE: Runtime target 'snort3-dev' is unbuildable, removing...
12:09:10 Missing or unbuildable dependency chain was: ['snort3-dev']
12:09:15 ERROR: Nothing RPROVIDES 'snort3' (but /mnt/jenkins/workspace/yocto-world-glibc/sources/meta-openembedded/meta-networking/recipes-connectivity/snort/snort3_git.bb RDEPENDS on or otherwise requires it)
12:09:15 No eligible RPROVIDERs exist for 'snort3'
12:09:15 NOTE: Runtime target 'snort3' is unbuildable, removing...
12:09:15 Missing or unbuildable dependency chain was: ['snort3']

```

File: meta-networking/recipes-connectivity/snort/snort3_git.bb (hunk 3 above)
Comment (RNGesus-exe) [reply meta-networking/recipes-connectivity/snort/snort3_git.bb:8]: sure, I'll try making this change, Also can I push all my changes in one single commit rather than making a commit for each recipe?

Comment (kraj) [reply meta-networking/recipes-connectivity/snort/snort3_git.bb:8]: Usually one commit per recipe is good

**Comment by RNGesus-exe** [comment]
Can you tell me how to reproduce the above errors on my end. I ended up fixing the bash dependency issue but I want to get the errors on my end so I can fix luajit + yocto build error

**Comment by kraj** [comment]
> @kraj Can you tell me how to reproduce the above errors on my end. I ended up isolating the bash dependent script into a seperate package but I want to get the errors on my end so I can fix luajit + yocto build error

I am building with clang/libc++, one way to reproduce it would be to use yoe distro https://github.com/yoedistro/yoe-distro

```
. ./envsetup.sh qemuriscv64
bitbake snort3
```

**Comment by RNGesus-exe** [comment]
I added the below code block to my snort3 recipe, is this what you want? I read the luajit recipe and apparently it is not supported for the following architectures yet.

`# mips64/ppc/ppc64/riscv64/riscv32 is not supported
COMPATIBLE_HOST:mipsarchn32 = "null"
COMPATIBLE_HOST:mipsarchn64 = "null"
COMPATIBLE_HOST:powerpc = "null"
COMPATIBLE_HOST:powerpc64 = "null"
COMPATIBLE_HOST:powerpc64le = "null"
COMPATIBLE_HOST:riscv64 = "null"
COMPATIBLE_HOST:riscv32 = "null"
`

**Comment by kraj** [comment]
Yes that would fix the compatiblity errror I am still worried about the build failures seen with clang and libc++

**Comment by RNGesus-exe** [comment]
Yup, that build error is still occurring for qemuarm, I am resolving it for now

**Comment by RNGesus-exe** [comment]
Update: I fixed the build error for qemuarm machine, by passing "-lm" compiler flag to the libdaq recipe

```
CFLAG += "-lm"
```

Now, libdaq is building but when building snort3, I get a new error something as below

```
/yoe/build/tmp/work/cortexa15t2hf-neon-yoe-linux-gnueabi/snort3/git/git/src/helpers/process.cc:191:56: error: use of undeclared identifier 'unw_strerror'; did you mean 'strerror'?
|   191 |         ssp.printf("unw_getcontext failed: %s (%d)\n", unw_strerror(ret), ret);
|       |                                                        ^~~~~~~~~~~~
|       |                                                        strerror
```

**Comment by RNGesus-exe** [comment]
> This error means a dependency on `libunwind` is missing.

I added it already, but it doesn't work. As meta-clang provides it's own libunwind

**Comment by kraj** [comment]
Can you set `TC_CXX_RUNTIME = "gnu"` and see if that helps ?

**Comment by kraj** [comment]
> > Can you set `TC_CXX_RUNTIME = "gnu"` and see if that helps ?
> 
> same error

Add it in meta-clang/conf/nonclangable.conf like below
```
TC_CXX_RUNTIME:pn-snort3:toolchain-clang = "gnu"
```

**Comment by RNGesus-exe** [comment]
> > > Can you set `TC_CXX_RUNTIME = "gnu"` and see if that helps ?
> > 
> > 
> > same error
> 
> Add it in meta-clang/conf/nonclangable.conf like below
> 
> ```
> TC_CXX_RUNTIME:pn-snort3:toolchain-clang = "gnu"
> ```

nope, still the same error [Yocto Build Error](https://pastebin.com/VKSGJj7C)

**Comment by RNGesus-exe** [comment]
I checked the environment `bitbake -e libunwind`, it is still being provided by libcxx

`FILE="/home/rngesusexe/Programming/yocto/yoe/sources/meta-clang/recipes-devtools/clang/libcxx_git.bb"`

**Comment by kraj** [comment]
Check for TC_CXX_RUNTIME what is it set to

**Comment by RNGesus-exe** [comment]
> Check for TC_CXX_RUNTIME what is it set to

I have pushed the new changes, but here the variable is set as follows
`TC_CXX_RUNTIME="llvm"`

**Comment by kraj** [comment]
> > Check for TC_CXX_RUNTIME what is it set to
> 
> I have pushed the new changes, but here the variable is set as following `TC_CXX_RUNTIME="llvm"`

Even when you set to gnu ? If so then that’s the issue

**Comment by RNGesus-exe** [comment]
> > > Check for TC_CXX_RUNTIME what is it set to
> > 
> > 
> > I have pushed the new changes, but here the variable is set as following `TC_CXX_RUNTIME="llvm"`
> 
> Even when you set to gnu ? If so then that’s the issue

Yup I added in meta-clang/conf/nonclangable.conf
`TC_CXX_RUNTIME:pn-snort3:toolchain-clang = "gnu"`

**Comment by RNGesus-exe** [comment]
@kraj is there any way to make `libunwind` from `poky/meta/recipes-support/libunwind/libunwind_1.6.2.bb` be used instead of the meta-clang one.

**Comment by kraj** [comment]
Please test https://github.com/openembedded/meta-openembedded/commit/8db172c5fe200053447c2df57a3d760a7efa451c

this should fix the problem.

**Comment by kraj** [comment]
Addition QA warnings

```
WARNING: snort3-3+git-r0 do_package_qa: QA Issue: File /usr/lib/pkgconfig/snort.pc in package snort3-dev contains reference to TMPDIR [buildpaths]
WARNING: snort3-3+git-r0 do_package_qa: QA Issue: File /usr/src/debug/snort3/3+git/src/js_norm/pdf_tokenizer.cc in package snort3-src contains reference to TMPDIR
File /usr/src/debug/snort3/3+git/src/js_norm/js_tokenizer.cc in package snort3-src contains reference to TMPDIR [buildpaths]

```

For once I have addressed them via

https://github.com/openembedded/meta-openembedded/commit/6803be9f6c1b83a0227254e19ec730c175448afb

However, I expect submitters to be able to handle these in general.

**Comment by RNGesus-exe** [comment]
> Please test [8db172c](https://github.com/openembedded/meta-openembedded/commit/8db172c5fe200053447c2df57a3d760a7efa451c)
> 
> this should fix the problem.

Tested, it is working

**Comment by kraj** [comment]
applied to master branch as of 7132d7b571

**Comment by shr-project** [comment]
In multilib builds I'm also seeing:
```
ERROR: lib32-snort3-3+git-r0 do_populate_sysroot: QA Issue: snort.pc failed sanity test (tmpdir) in path lib32-snort3/3+git/sysroot-destdir/usr/lib/pkgconfig [pkgconfig]
```

And the ${SRCPV} shouldn't be in PV, see:
https://git.openembedded.org/meta-openembedded/commit/?id=21f956598ddb1bb08855fbc9a3293acf957154d1

**Comment by shr-project** [comment]
I'll send fixes for both issues to ML shortly.

**Comment by RNGesus-exe** [comment]
@shr-project How can I reproduce this error


> In multilib builds I'm also seeing:
> 
> ```
> ERROR: lib32-snort3-3+git-r0 do_populate_sysroot: QA Issue: snort.pc failed sanity test (tmpdir) in path lib32-snort3/3+git/sysroot-destdir/usr/lib/pkgconfig [pkgconfig]
> ```


moreover can you kindly give some info on how to resolve this issue.

**Comment by shr-project** [comment]
See the fixes I've sent to ML and were already merged (see https://git.openembedded.org/meta-openembedded/log/?qt=grep&q=snort3)
